File: source-code/interfacing-c-c++-fortran/Pybind11/Convolution/src/test_convolution.cpp
#include <convolution.h>
#include <iostream>

int main() {
    // Create a 10x10 image
    Matrix image(10, 10);
    for (int i = 0; i < image.rows(); ++i) {
        for (int j = 0; j < image.cols(); ++j) {
            image(i, j) = i*image.cols() + j;
        }
    }

    // Print the image
    std::cout << image << std::endl;

    // Create a 3x3 kernel
    Matrix kernel(3, 3);
    for (int i = 0; i < kernel.rows(); ++i) {
        for (int j = 0; j < kernel.cols(); ++j) {
            kernel(i, j) = 1.0/(kernel.rows()*kernel.cols());
        }
    }

    // Print the kernel
    std::cout << kernel << std::endl;

    // Create a convolution object
    auto new_image = convolve(image, kernel);

    // Print the result
    std::cout << new_image << std::endl;

    return 0;
}

File: source-code/interfacing-c-c++-fortran/Pybind11/Convolution/test_convolution.py
#!/usr/bin/env python

import argparse
from build.src.bindings.convolve import convolve
import numpy as np
import timeit


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--img_size', type=int, default=10, help='number of rows in the image')
    parser.add_argument('--kernel_size', type=int, default=3, help='number of rows in the kernel')
    parser.add_argument('--num_iter', type=int, default=1, help='number of iterations to run')
    args = parser.parse_args()

    # Create input image and kernel
    input_image = np.random.uniform(0.0, 1.0, size=(args.img_size, args.img_size))
    kernel = np.random.rand(args.kernel_size, args.kernel_size)
    result = np.empty((args.img_size + args.kernel_size - 1, args.img_size + args.kernel_size - 1))
    time = timeit.timeit(lambda: convolve(input_image, kernel, result), number=args.num_iter)
    print(f'Image size: {args.img_size}x{args.img_size}')
    print(f'Kernel size: {args.kernel_size}x{args.kernel_size}')
    print(f'Time: {time/args.num_iter:.6f} seconds')
    print(f'Sum: {np.sum(result)}')

if __name__ == '__main__':
    main()
